### PR TITLE
plan: Fix the empty result of join

### PR DESCRIPTION
**Pull request by zz-jason**
## Brief Intro 
Let's say we have the following table:
```sql
drop table if exists t;
create table t (a bigint);
insert into t values (1);
```

Before this PR:
```sql
TiDB(localhost:4000) > select t2.a, t1.a from t t1 inner join (select "1" as a) t2 on t2.a = t1.a;
Empty set (0.00 sec)
```

After this PR:
```sql
TiDB(localhost:4000) > select t2.a, t1.a from t t1 inner join (select "1" as a) t2 on t2.a = t1.a;
+------+------+
| a    | a    |
+------+------+
| 1    |    1 |
+------+------+
1 row in set (0.00 sec)
```
## What is the root cause of this issue

The execution plan of this sql is:
```sql
TiDB(localhost:4000) > desc select t2.a, t1.a from t t1 inner join (select "1" as a) t2 on t2.a = t1.a;
+----------------+----------------+-----------------------------+------+------------------------------------------------------------------+----------+
| id             | parents        | children                    | task | operator info                                                    | count    |
+----------------+----------------+-----------------------------+------+------------------------------------------------------------------+----------+
| TableScan_11   |                |                             | cop  | table:t1, range:[-inf,+inf], keep order:false                    | 10000.00 |
| TableReader_12 | Projection_10  |                             | root | data:TableScan_11                                                | 10000.00 |
| Projection_10  | HashLeftJoin_8 | TableReader_12              | root | t1.a, cast(t1.a)                                                 | 10000.00 |
| TableDual_14   | Projection_13  |                             | root | rows:1                                                           | 1.00     |
| Projection_13  | HashLeftJoin_8 | TableDual_14                | root | 1, cast(1)                                                       | 1.00     |
| HashLeftJoin_8 | Projection_7   | Projection_10,Projection_13 | root | inner join, inner:Projection_13, equal:[eq(cast(t1.a), cast(1))] | 1.25     |
| Projection_7   |                | HashLeftJoin_8              | root | t2.a, t1.a                                                       | 1.25     |
+----------------+----------------+-----------------------------+------+------------------------------------------------------------------+----------+
7 rows in set (0.00 sec)
```

NOTE:
For `Projection_13`, the expressions are `1` and `cast(1)`:
- `1` is built in: https://github.com/pingcap/tidb/pull/6669/files#diff-638d8be162d442ef3a5c423ff0c735d4R580
- `cast(1)` is built in: https://github.com/pingcap/tidb/blob/master/plan/rule_predicate_push_down.go#L200

before this PR, these two expressions have the same `FromID`, and `Position` in the schema of `Projection_13`, which cause the wrong join key chosen, and further cause the empty join result.

### CODE REVIEW

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@winoros PTAL

File: plan/logical_plan_builder.go [hunk 1]
```diff
@@ -577,7 +577,7 @@ func (b *planBuilder) buildProjection(p LogicalPlan, fields []*ast.SelectField,
 		p = np
 		proj.Exprs = append(proj.Exprs, newExpr)
 
-		col := b.buildProjectionField(proj.id, schema.Len()+1, field, newExpr)
+		col := b.buildProjectionField(proj.id, i, field, newExpr)
```
Comment (winoros) [inline plan/logical_plan_builder.go]: Why this change can take effect?

Comment (zz-jason) [reply plan/logical_plan_builder.go]: see the description for more details.

Comment (winoros) [reply plan/logical_plan_builder.go]: why not just use `schema.Len()`

Comment (zz-jason) [reply plan/logical_plan_builder.go]: use `i` is shorter

Comment (zz-jason) [reply plan/logical_plan_builder.go]: and more specific
